social_django/views.py

```diff
@@ -39,21 +43,82 @@ def disconnect(request, backend, association_id=None):
+            expiration.seconds + (expiration.days * 86400)
+        ) if expiration else DEFAULT_SESSION_TIMEOUT
+
+        # Check to see if the backend set a value as a maximum length
+        # that a session may be; if they did, then we should use the minimum
+        # of that and the received session expiration time, if any, to
+        # set the session length.
+        if received_expiration_time is None and max_session_length is None:
+            # We neither received an expiration length, nor have a maximum
+            # session length. Use the platform default.
+            session_expiry = DEFAULT_SESSION_TIMEOUT
+        elif max_session_length is not None and received_expiration_time is None:
+            # We only have a maximum session length; use that.
+            session_expiry = max_session_length
+        elif received_expiration_time is not None and max_session_length is None:
+            # We only have an expiration time received by the backend
+            # from the provider, with no set maximum. Use that.
+            session_expiry = received_expiration_time
+        else:
+            # We received an expiration time from the backend, and we also
+            # have a set maximum session length. Use the smaller of the two.
+            session_expiry = min(received_expiration_time, max_session_length)
+    else:
+        # If there's an explicitly-set maximum session length, use that
+        # even if we don't want to retrieve session expiry times from
+        # the backend. If there isn't, then use the platform default.
+        session_expiry = max_session_length or DEFAULT_SESSION_TIMEOUT
+
+    return session_expiry
+
+
 def _do_login(backend, user, social_user):
     user.backend = '{0}.{1}'.format(backend.__module__,
                                   backend.__class__.__name__)
+    # Get these details early to avoid any issues involved in the
+    # session switch that happens when we call login().
+    enable_session_expiration = backend.setting('SESSION_EXPIRATION', False)
+    max_session_length_setting = backend.setting('MAX_SESSION_LENGTH', None)
+
+    # Log the user in, creating a new session.
     login(backend.strategy.request, user)
-    if backend.setting('SESSION_EXPIRATION', False):
-        # Set session expiration date if present and enabled
-        # by setting. Use last social-auth instance for current
-        # provider, users can associate several accounts with
-        # a same provider.
-        expiration = social_user.expiration_datetime()
-        if expiration:
-            try:
-                backend.strategy.request.session.set_expiry(
-                    expiration.seconds + expiration.days * 86400
-                )
-            except OverflowError:
-                # Handle django time zone overflow
-                backend.strategy.request.session.set_expiry(None)
+
+    # Make sure that the max_session_length value is either an integer or None.
```

Comment:
Good call! We'd specifically be concerned about cases where the value might be set to a string that can't be interpreted as an int, or something like that.